### PR TITLE
feat: run snapshot tests dynamically in unimarkup-core crate

### LINKED ISSUE #100: [REQ] Dynamic test implementation
In #97 dynamic testing is used to generate snapshot test cases from `*.yml` files available. How that works is, the directory where snapshot tests are stored is walked through recursively, test files are parsed and test cases are created. Using a custom test harness, we are even able to get each separate test case listed when testing with `cargo t`. 

This should be ported to `unimarkup-commons::test_runner` so that we can use it in other crates as well.

### PR DESCRIPTION
**Pull request by nfejzic**
This PR adds dynamic testing of snapshots in `unimarkup_core` crate. 

By dynamic is meant that all spec files are automatically collected, parsed in order to generate test cases, and these test cases are run. Snapshot files are automatically created, and each test case (where each spec yaml file can contain multiple tests) is reported in `cargo test` output with corresponding name (and module path). 

Downside of using custom harness is that using `#[test]` attribute to automatically test functions does not work (as far as I know). Any custom function must be manually collected. See `test_fn!` macro and how it's used. 

Closes #100.

### CODE REVIEW

File: core/tests/blocks.rs
Comment (mhatzl) [inline core/tests/blocks.rs:1]: This tests all markup files inside the `/markup` folder, not just `blocks`.
Maybe rename it `runner.rs` and move it to `/tests/snapshot` instead?

File: core/tests/blocks.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+use unimarkup_commons::test_runner::{self, test_file};
+use unimarkup_core::Unimarkup;
+
+pub fn test_block_snapshots() -> Vec<libtest_mimic::Trial> {
```
Comment (mhatzl) [inline core/tests/blocks.rs:4]:
```suggestion
pub fn test_core_snapshots() -> Vec<libtest_mimic::Trial> {
```

See remark on the file.

File: core/tests/blocks.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        };
+
+        test_runs.push(libtest_mimic::Trial::test(test_name, test_run));
+    }
+
+    test_runs
+}
+
+fn run_test_case(case: test_runner::test_file::TestCase) {
+    test_runner::spec_test::assert_um_spec(
```
Comment (mhatzl) [inline core/tests/blocks.rs:35]: This asserts that out implementation returned the correct output (currently only HTML).
It is not a snapshot test.
For snapshot testing, you need to add `test_parser_snap!()` from `unimarkup-commons`.

**Review by mhatzl — COMMENTED** [review]
Thanks for this PR.

The `blocks.rs` file in `core` seems to be missing the snapshot test runner.
At the moment, only spec tests are run if I am correct.

Everything else looks ready to be merged.

File: Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,7 @@ members = [
     "parser",
     "render",
 ]
+resolver = "2"
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply Cargo.toml:10]: Me neither, but `cargo` kept bugging me until I changed it 🤣

File: core/tests/blocks.rs
Comment (nfejzic) [reply core/tests/blocks.rs:1]: Moved to `tests/runner/mod.rs`.

File: core/tests/blocks.rs (hunk 2 above)
Comment (nfejzic) [reply core/tests/blocks.rs]: I now collect snapshot tests as well. Also the test output now includes that bit of info for each test. 

For example: 

```
test unimarkup_core::runner::spec::simple-paragraph       ... ok
test unimarkup_core::runner::spec::paragraph-with-inlines ... ok
test unimarkup_core::runner::snap::paragraph-with-inlines ... ok
test unimarkup_core::runner::snap::simple-paragraph       ... ok
```

File: core/tests/runner/mod.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        "markup",
+    );
+
+    let mut test_runs = Vec::with_capacity(test_cases.len());
+
+    for case in test_cases {
+        let spec_test_name = format!("{}::spec::{}", module_path!(), case.test.name.as_str());
+        let snap_test_name = format!("{}::snap::{}", module_path!(), case.test.name.as_str());
+
+        let case2 = case.clone();
```
Comment (mhatzl) [inline core/tests/runner/mod.rs:23]:
```suggestion
        let cloned_case = case.clone();
```

Because `case2` sounds like it would be a different case.

**Review by mhatzl — COMMENTED** [review]
Looks good now.
Just not sure about `case2` as variable name.

But I can change that myself once I merge #106 with these changes.
There will be some merge conflicts anyways, because of the Scanner change.

File: core/tests/runner/mod.rs (hunk 4 above)
Comment (nfejzic) [reply core/tests/runner/mod.rs]: You're right, I renamed it now!

**Review by mhatzl — APPROVED** [review]
thanks 😊
looks good now 👍

**Pull request merged**